### PR TITLE
Duplicate widgets

### PR DESCRIPTION
**Pull request by Spoffy**
## Context

If a user wants to duplicate a widget, the only option currently is to do it manually. 

## Proposed solution

This adds a 'duplicate widget' option to the widget, allowing the widget to be copied to any page.

It reuses the duplicatePage logic for the most part, with some changes to facilitate it working with individual widgets.

## Related issues

#1644 

## Has this been tested?

- [x] 👍 yes, I added tests to the test suite
- [x] 💭 no, because this PR is a draft and still needs work
- [ ] 🙅 no, because this is not relevant here
- [ ] 🙋 no, because I need help 

## Screenshots / Screencasts

TBC

### CODE REVIEW

File: app/client/components/duplicateWidget.ts [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+import flatten from 'lodash/flatten';
+import forEach from 'lodash/forEach';
+import {testId} from 'app/client/ui2018/cssVars';
+import {logTelemetryEvent} from 'app/client/lib/telemetry';
+import sortBy from 'lodash/sortBy';
+import {fromPairs} from 'lodash';
+
+const t = makeT('duplicateWidget');
+
+// Duplicate page with pageId. Starts by prompting user for a new name.
```
Comment (berhalak) [inline app/client/components/duplicateWidget.ts]: Out of date

File: app/client/components/duplicateWidget.ts [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+    return {
+      label: `${view.name.peek()}${suffix}`,
+      value: view.getRowId(),
+      isActivePage: isActivePage,
+    };
+  });
+
+  pageSelectOptions.push({ label: 'Create new page', value: 0, isActivePage: false });
+
+  // Logically this should never happen, as a Grist doc without pages should be impossible.
```
Comment (berhalak) [inline app/client/components/duplicateWidget.ts]: It is impossible because you've just added an item above :)

File: app/client/components/duplicateWidget.ts [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+export async function duplicateWidgets(gristDoc: GristDoc, srcViewSectionIds: number[], destViewId: number) {
+  const allViewSectionModels = gristDoc.docModel.viewSections.rowModels;
+  const validWidgetIds = srcViewSectionIds.filter(id => allViewSectionModels[id]);
+
+  // Generally this shouldn't happen, but it catches a theoretically possible edge case.
+  if (validWidgetIds.length === 0) {
+    throw new Error("Unable to duplicate widgets as no valid source widget IDs were provided");
+  }
+
+  const sourceViewSections = validWidgetIds.map(id => allViewSectionModels[id]);
```
Comment (berhalak) [inline app/client/components/duplicateWidget.ts]: Maybe you can merge this with the line in 71, to run this filter only once (and the `validWidgetIds` is not needed afterward)

File: app/client/components/duplicateWidget.ts [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+  // occurs before add.
+  const results = await gristDoc.docData.sendActions([
+    addAction,
+  ]);
+
+  const newFieldIds: number[] = results[0];
+  return fromPairs(srcViewFieldIds.map((srcId, index) => [srcId, newFieldIds[index]]));
+}
+
+async function listAllViewFields(viewSections: ViewSectionRec[]) {
```
Comment (berhalak) [inline app/client/components/duplicateWidget.ts]: Why this is async?

File: app/client/components/duplicateWidget.ts [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+      const {
+        duplicatedViewSections,
+        viewRef,
+        viewSectionIdMap
+      } = await createNewViewSections(gristDoc, sourceViewSections, destViewId);
+      resolvedDestViewId = viewRef;
+
+      const defaultViewFieldIds = await listAllViewFields(duplicatedViewSections.map(pair => pair.destViewSection));
+
+      const viewFieldsIdMap = await copyOriginalViewFields(gristDoc, duplicatedViewSections);
```
Comment (berhalak) [inline app/client/components/duplicateWidget.ts:109]: I'm lost by here. Can you add some narration up to this moment?

File: app/client/components/duplicateWidget.ts [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+    t("Duplicate widgets"),
+    async () => {
+      const {
+        duplicatedViewSections,
+        viewRef,
+        viewSectionIdMap
+      } = await createNewViewSections(gristDoc, sourceViewSections, destViewId);
+      resolvedDestViewId = viewRef;
+
+      const defaultViewFieldIds = await listAllViewFields(duplicatedViewSections.map(pair => pair.destViewSection));
```
Comment (berhalak) [inline app/client/components/duplicateWidget.ts]: Can you move it closer to where it is used?

File: app/client/components/duplicateWidget.ts [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+
+      const viewFieldsIdMap = await copyOriginalViewFields(gristDoc, duplicatedViewSections);
+
+      // update layout spec
+      let layoutSpecUpdatePromise = Promise.resolve();
+      // If we're creating a new page, we should copy the widget layout over.
```
Comment (berhalak) [inline app/client/components/duplicateWidget.ts]: Why?

File: app/client/components/duplicateWidget.ts [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+      if (isNewView) {
+          const newLayoutSpec = patchLayoutSpec(sourceView.layoutSpecObj.peek(), viewSectionIdMap);
+          layoutSpecUpdatePromise = gristDoc.docData.sendAction(
+            ['UpdateRecord', '_grist_Views', resolvedDestViewId, { layoutSpec: JSON.stringify(newLayoutSpec)}]
+          );
+      }
+      await Promise.all([
+        layoutSpecUpdatePromise,
+        updateViewSections(gristDoc, duplicatedViewSections, viewFieldsIdMap, viewSectionIdMap),
```
Comment (berhalak) [inline app/client/components/duplicateWidget.ts:122]: Can you change the name, or split it to more focused methods?

File: app/client/components/duplicateWidget.ts [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+}
+
+/**
+ * Update all of destViewSections with srcViewSections, use fieldsMap to patch the section layout
+ * (for detail/cardlist sections), use viewSectionMap to patch the sections ids for linking.
+ */
+async function updateViewSections(gristDoc: GristDoc, duplicatedViewSections: DuplicatedViewSection[],
+                                  fieldsMap: {[id: number]: number}, viewSectionMap: {[id: number]: number}) {
+
+  // collect all the records for the src view sections
```
Comment (berhalak) [inline app/client/components/duplicateWidget.ts]: src or dest?

File: app/client/components/duplicateWidget.ts [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+  const destRowIds: number[] = [];
+  const records: RowRecord[] = [];
+  for (const { srcViewSection, destViewSection } of duplicatedViewSections) {
+    const viewSectionLayoutSpec =
+      srcViewSection.parentKey.peek() === 'form'
+          ? cleanFormLayoutSpec(srcViewSection.layoutSpecObj.peek(), fieldsMap)
+          : patchLayoutSpec(srcViewSection.layoutSpecObj.peek(), fieldsMap);
+    const record = gristDoc.docData.getMetaTable('_grist_Views_section').getRecord(srcViewSection.getRowId())!;
+
+    const isNewView = srcViewSection.view.peek().id.peek() != destViewSection.view.peek().id.peek();
```
Comment (berhalak) [inline app/client/components/duplicateWidget.ts:191]: - I don't get this fragment, can you add a comment. I don't get why you are checking isNewView (or what for).
- You already have the record, you don't need to check the model.

File: app/client/components/duplicateWidget.ts [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+      layoutSpec: JSON.stringify(viewSectionLayoutSpec),
+      linkSrcSectionRef: linkRef ?? false,
+      shareOptions: '',
+    });
+  }
+
+  // transpose data
+  const sectionsInfo = getColValues(records);
+
+  // ditch column parentId
```
Comment (berhalak) [inline app/client/components/duplicateWidget.ts]: This comment doesn't help :)

File: app/client/components/duplicateWidget.ts [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+  delete sectionsInfo.parentId;
+
+  // send action
```
Comment (berhalak) [inline app/client/components/duplicateWidget.ts]: Neither does this :)

File: app/client/components/duplicateWidget.ts [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+            ['UpdateRecord', '_grist_Views', resolvedDestViewId, { layoutSpec: JSON.stringify(newLayoutSpec)}]
+          );
+      }
+      await Promise.all([
+        layoutSpecUpdatePromise,
+        updateViewSections(gristDoc, duplicatedViewSections, viewFieldsIdMap, viewSectionIdMap),
+        copyFilters(gristDoc, sourceViewSections, viewSectionIdMap)
+      ]);
+
+      // Remove default view fields at the end - removing them earlier means they're still referenced in
```
Comment (berhalak) [inline app/client/components/duplicateWidget.ts]: Can you explain more?

**Review by berhalak — CHANGES_REQUESTED** [review]
When I duplicate a section it produces two bundles, not one. So to undo I need to undo twice.

File: app/client/components/ViewLayout.ts [hunk 14]
```diff
@@ -43,6 +43,7 @@ import {
 } from 'grainjs';
 import * as ko from 'knockout';
 import debounce from 'lodash/debounce';
+import {buildDuplicateWidgetModal} from 'app/client/components/duplicateWidget';
```
Comment (berhalak) [inline app/client/components/ViewLayout.ts]: Sort order

File: app/client/ui/ViewLayoutMenu.ts [hunk 15]
```diff
@@ -57,6 +57,9 @@ export function makeViewLayoutMenu(viewSection: ViewSectionRec, isReadonly: bool
     (gristDoc.externalSectionId.get() === viewSection.getRowId()) ||
     (gristDoc.maximizedSectionId.get() === viewSection.getRowId());
 
+  const dontDuplicateSection = (use: UseCB) =>
+    use(viewSection.isRaw) || use(viewSection.isVirtual);
```
Comment (berhalak) [inline app/client/ui/ViewLayoutMenu.ts]: Also `isReadonly`.

File: test/nbrowser/duplicateWidget.ts [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+    it('can duplicate the widget', async function() {
+      // Initial setup of the base widget
+      const allSections = await gu.getSectionTitles();
+      await gu.selectSectionByTitle(allSections[0]);
+
+      await gu.renameSection(allSections[0], 'Widget 1');
+
+      await gu.getCell('A', 1).click();
+      await driver.sendKeys(Key.ENTER + testCellContent + Key.ENTER);
+      await driver.sendKeys(Key.DOWN);
```
Comment (berhalak) [inline test/nbrowser/duplicateWidget.ts]: This is flaky, and the Key.DOWN is not needed, Key.Enter already moves to the next row

```suggestion
      await gu.waitAppFocus(); // this is needed, focus is catched asynchronously (by focus layer and timeout)
      await gu.enterCell(testCellContent);
      await gu.enterCell(testCellContent2);
```

File: test/nbrowser/duplicateWidget.ts [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+    it('can duplicate a widget with selectby and style rules', async function() {
+      await gu.changeWidget('Table');
+
+      await gu.addInitialStyleRule();
+      await gu.openStyleRuleFormula(0);
+      await driver.sendKeys(testFormula + Key.ENTER);
+      await gu.selectBy('Widget 1');
+
+      await gu.duplicateWidget('Widget 2');
+      // Reduces flakiness by waiting until duplication is finished and it's okay to rename.
```
Comment (berhalak) [inline test/nbrowser/duplicateWidget.ts:110]: Maybe it is caused by this double undo issue?

File: test/nbrowser/duplicateWidget.ts [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+  const testCellContent = "Rubber duck";
+  const testCellContent2 = "Plastic boat";
+
+  before(async function() {
+    const session = await gu.session().teamSite.login();
+    await session.tempNewDoc(cleanup, 'DuplicateWidget.grist');
+  });
+
+  describe("duplicating a widget on the same page", async function() {
+    it('can duplicate the widget', async function() {
```
Comment (berhalak) [inline test/nbrowser/duplicateWidget.ts:20]: Can you add a test that removes one field or changes its name and makes sure that orginal section is not changed.

Also I'm missing tests for other widgets (charts, forms etc), and a test that duplicates a widget with hidden (by ACL) column [this is tricky].

File: app/client/components/duplicateWidget.ts (hunk 2 above)
Comment (Spoffy) [reply app/client/components/duplicateWidget.ts]: :rofl:, great catch, thanks  :)

File: app/client/components/duplicateWidget.ts (hunk 4 above)
Comment (Spoffy) [reply app/client/components/duplicateWidget.ts]: Probably leftover from a refactor, nice find :)

File: app/client/components/duplicateWidget.ts (hunk 5 above)
Comment (Spoffy) [reply app/client/components/duplicateWidget.ts:109]: I've added some comments that better explain what's happening and why. 

It might be possible to refactor a chunk of this to make it more intuitive just from the code, but I'm wary of touching this again now it's working.

File: app/client/components/duplicateWidget.ts (hunk 6 above)
Comment (Spoffy) [reply app/client/components/duplicateWidget.ts]: Unfortunately I can't. The deletion has to happen at the end, this has to happen straight after view section creation, before the new view fields are added. :disappointed:

File: app/client/components/duplicateWidget.ts (hunk 7 above)
Comment (Spoffy) [reply app/client/components/duplicateWidget.ts]: The duplicatePage function uses this logic - although actually, I guess I could refactor it so this specific bit of logic lives in duplicate page again?

File: app/client/components/duplicateWidget.ts (hunk 8 above)
Comment (Spoffy) [reply app/client/components/duplicateWidget.ts:122]: Potentially - I'd like to. A lot of this was copied straight from the duplicate page logic, and it's pretty optimised to avoid sending too many actions - which makes it tricky to refactor.

File: app/client/components/duplicateWidget.ts (hunk 9 above)
Comment (Spoffy) [reply app/client/components/duplicateWidget.ts]: Old comment from when this was duplicatePage, removed it.

File: app/client/components/duplicateWidget.ts (hunk 10 above)
Comment (Spoffy) [reply app/client/components/duplicateWidget.ts:191]: Yeah, will try and clean this up a bit.

The core idea here:
- If the copied widget is going on a new page, and we're copying multiple widgets, we want to link the new widget A to the _copy_ of the widget it was originally linked to. 
- If the copied widget is going on a new page, and we're copying a single widget, linkRef should be undefined.
- If the copied widget is staying on the same page, the copied widget and the source widget should have the same linkRef. (I.e - their selections are both controlled by the same widget)

I'm having difficulty explaining this clearly, even in normal language :laughing:

File: app/client/components/duplicateWidget.ts (hunk 12 above)
Comment (Spoffy) [reply app/client/components/duplicateWidget.ts]: Feeling like I should have spent more time cleaning up the original duplicatePage code :D

File: app/client/components/duplicateWidget.ts (hunk 13 above)
Comment (Spoffy) [reply app/client/components/duplicateWidget.ts]: I've expanded this substantially - let me know if it's now _too_ much :laughing:

File: app/client/ui/ViewLayoutMenu.ts (hunk 15 above)
Comment (Spoffy) [reply app/client/ui/ViewLayoutMenu.ts]: Wasn't sure if duplicating something readOnly would be acceptable

File: test/nbrowser/duplicateWidget.ts (hunk 16 above)
Comment (Spoffy) [reply test/nbrowser/duplicateWidget.ts]: Thanks :)

File: test/nbrowser/duplicateWidget.ts (hunk 17 above)
Comment (Spoffy) [reply test/nbrowser/duplicateWidget.ts:110]: Not a clue, but I'll re-test after that's fixed :)

File: test/nbrowser/duplicateWidget.ts (hunk 18 above)
Comment (Spoffy) [reply test/nbrowser/duplicateWidget.ts:20]: There's some tricky ones in there for sure.  Will see what I can do.

File: app/client/components/duplicateWidget.ts (hunk 1 above)
Comment (Spoffy) [reply app/client/components/duplicateWidget.ts]: Removed!

File: app/client/components/duplicateWidget.ts (hunk 3 above)
Comment (Spoffy) [reply app/client/components/duplicateWidget.ts]: Done - double check you're happy with it? :)

File: test/nbrowser/duplicateWidget.ts (hunk 16 above)
Comment (berhalak) [reply test/nbrowser/duplicateWidget.ts]: But have you seen my suggestion? Your fix with replacing Key.Down line doesn't help here (from my understanding).
- When we click a cell, it doesn't mean that it has a focus, focus still for a brief moment can be somewhere else, we need to wait for it a bit. This changed recently, as somethings became more async and browser became much faster
- We need to wait for the server after every Key.Enter.

**Comment by berhalak** [comment]
Looks ok, only 2 issues:
1. The duplication is not bundled, so it requires 2 undos to go back
2. The test code in in `duplicateWidget` is flaky.

File: test/nbrowser/duplicateWidget.ts (hunk 17 above)
Comment (Spoffy) [reply test/nbrowser/duplicateWidget.ts:110]: This should be unrelated to the double undo - that was caused by an empty action being sent to the server, breaking the bundle.

**Comment by Spoffy** [comment]
I have read the CLA Document and I hereby sign the CLA

File: test/nbrowser/duplicateWidget.ts (hunk 18 above)
Comment (Spoffy) [reply test/nbrowser/duplicateWidget.ts:20]: Added:
- Tests for forms
- Tests for hiding a field on a duplicating widget, and making sure it doesn't hide one on the original

File: test/nbrowser/duplicateWidget.ts (hunk 16 above)
Comment (Spoffy) [reply test/nbrowser/duplicateWidget.ts]: I realised I didn't understand your original suggestion correctly! Does this look better now?
